Add inputs and outputs in Triton Kernel FX Graph segment

shengfukevin opened the pull request:
Summary: Add inputs and outputs in Triton Kernel FX Graph segment

The FX graph segment in Triton kernel does not include the input tensors and return tensors, for example
Python code:
```
  @torchdynamo.optimize("inductor")
  def fn(a, b, c):
      x = torch.nn.functional.linear(a, b)
      x = x.sin()
      x = x.t() + c * 2
      return x
```

```
# %sin : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.sin.default](args = (%mm,), kwargs = {})
# %permute_1 : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%sin, [1, 0]), kwargs = {})
# %mul : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg2_1, 2), kwargs = {})
# %add : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%permute_1, %mul), kwargs = {})

```
The fix is to add the input and output tensors into FX graph segment

```
# %mm : Tensor "f32[4, 4][4, 1]cuda:0" = PlaceHolder[target=mm]
# %arg2_1 : Tensor "f32[4, 4][4, 1]cuda:0" = PlaceHolder[target=arg2_1]
# %sin : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.sin.default](args = (%mm,), kwargs = {})
# %permute_1 : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%sin, [1, 0]), kwargs = {})
# %mul : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg2_1, 2), kwargs = {})
# %add : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%permute_1, %mul), kwargs = {})
# return %add
```

Differential Revision: D78131358

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

ezyang reviewed (dismissed):
Please show before and after output

jansel commented on torch/_inductor/utils.py:
```diff
@@ -766,11 +767,65 @@ def get_kernel_metadata(
+                        output_name, _ = get_buffer_info(buffer)
+
+                        all_writes.append("%" + output_name)
+        except Exception:
+            pass
```
What is this?  We should not silently ignore errors.

shengfukevin replied:
I saw a few errors when running PyTorch CI, not sure how to fix it. Add try catch here to avoid the crash. I printed out the exception as a warning.

jansel replied:
We should fix the root cause of these issues, not hide them.

jansel reviewed (changes requested):
Add a test?

shengfukevin commented:
@jansel, I updated this PR based on your comments, can you review it again? Thanks

jansel commented on test/profiler/test_execution_trace.py:
```diff
@@ -469,6 +469,103 @@ def fn(a, b, c):
+                        else:
+                            fx_graph_found = False
+
+                    if len(fx_graph) > 0:
```
When is this false?  What if it is false for all files?

jansel commented on torch/_inductor/utils.py:
```diff
@@ -758,11 +759,72 @@ def get_kernel_metadata(
+                        output_name, _ = get_buffer_info(buffer)
+
+                        all_writes.append("%" + output_name)
+        except Exception as e:
```
I repeat my prior comment.  We should not blindly catch all errors, we should fix and fix the root cause of the errors.

jansel reviewed (changes requested):
Lint errors

shengfukevin commented:
@jansel, I removed the exception catch block in utils.py, and fixed all failed unit tests. Please review it. Thanks

jansel commented:
@shengfukevin lints are still failing and you didn't answer my other comment

shengfukevin commented:
> @shengfukevin lints are still failing and you didn't answer my other comment

The lint error is for doc string of "define_kernel_metadata". It is a existing issue, the whole file does not have doc string.

shengfukevin committed:
Add inputs and outputs in Triton Kernel FX Graph segment (#158174)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/158174

Add inputs and outputs in Triton Kernel FX Graph segment

The FX graph segment in Triton kernel does not include the input tensors and return tensors, for example

Python code:
    torchdynamo.optimize("inductor")
    def fn(a, b, c):
        x = torch.nn.functional.linear(a, b)
        x = x.sin()
        x = x.t() + c * 2
        return x

`#   %sin : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.sin.default](args = (%mm,), kwargs = {})`
`#   %permute_1 : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%sin, [1, 0]), kwargs = {})`
`#   %mul : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg2_1, 2), kwargs = {})`
`#   %add : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%permute_1, %mul), kwargs = {})`

The fix is to add the input and output tensors into FX graph segment

`#   %mm : Tensor "f32[4, 4][4, 1]cuda:0" = PlaceHolder[target=mm]`
`#   %arg2_1 : Tensor "f32[4, 4][4, 1]cuda:0" = PlaceHolder[target=arg2_1]`
`#   %sin : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.sin.default](args = (%mm,), kwargs = {})`
`#   %permute_1 : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%sin, [1, 0]), kwargs = {})`
`#   %mul : "f32[4, 4][4, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg2_1, 2), kwargs = {})`
`#   %add : "f32[4, 4][1, 4]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%permute_1, %mul), kwargs = {})`
`#   return %add`

Test Plan:
buck2 run mode/opt durin/api:test_remote_fx_graph_cache
buck2 run mode/opt durin/api:query_remote_fx_graph_cache -- --remote-cache-key xxxx

Rollback Plan:

Differential Revision: D78131358

shengfukevin commented:
@pytorchbot merge

shengfukevin commented:
@ezyang, can you review this PR?
